src/sage/rings/species.py

```diff
@@ -0,0 +1,1751 @@
+
+        If `G = (X, a)`, then `X` should be a finite set and `a` a transitive
+        action of `G` on `X`.
+
+        TESTS::
+
+            sage: P = PolynomialSpecies(ZZ, ["X", "Y"])
+            sage: M = P._indices
+            sage: M(CyclicPermutationGroup(4), {1: [1,2], 2: [3,4]})
+            Traceback (most recent call last):
+            ...
+            ValueError: All elements of orbit (1, 2, 3, 4) must have the same sort
+        """
+        if parent(G) == self:
+            if pi is not None:
+                raise ValueError("cannot reassign sorts to a molecular species")
+            return G
+        if isinstance(G, PermutationGroup_generic):
+            if pi is None:
+                if self._k == 1:
+                    pi = {1: G.domain()}
+                else:
+                    raise ValueError("the assignment of sorts to the domain elements must be provided")
+            domain_partition = G.disjoint_direct_product_decomposition()
+            elm = self.one()
+            for part in domain_partition:
+                elm *= self.gen(self._project(G, pi, part))
+            return elm
+        # Assume G is a tuple (X, a)
+        X, a = G
+        if pi is None:
+            if self._k == 1:
+                pi = {1: X}
+            else:
+                raise ValueError("the assignment of sorts to the domain elements must be provided")
+        L = [None for _ in range(self._k)]
+        for k, v in pi.items():
+            L[k - 1] = list(v)
+        # Create group
+        # TODO: Is this correct?
+        S = SymmetricGroup(list(chain.from_iterable(L))).young_subgroup([len(v) for v in L])
+        H = PermutationGroup(S.gens(), action=a, domain=X)
+        if len(H.orbits()) > 1:
+            # Then it is not transitive
+            raise ValueError("Action is not transitive")
+        stabG = PermutationGroup([g for g in S.gens() if a(g, H.orbits()[0][0]) == H.orbits()[0][0]], domain=S.domain())
+        return self(stabG, pi)
+
+    @cached_method
+    def one(self):
+        r"""
+        Return the one of this monoid.
+
+        EXAMPLES::
+
+            sage: P = PolynomialSpecies(ZZ, "X, Y")
+            sage: P._indices.one()
+            1
+        """
+        elm = super().one()
+        elm._group = SymmetricGroup(0)
+        elm._dompart = tuple()
+        elm._mc = tuple(0 for _ in range(self._k))
+        elm._tc = 0
+        return elm
+
+    def gen(self, x):
+        r"""
+        Create the molecular species from an atomic species.
+
+        EXAMPLES::
+
+            sage: P = PolynomialSpecies(ZZ, "X, Y")
+            sage: At = AtomicSpecies("X, Y")
+            sage: G = PermutationGroup([[(1,2),(3,4)]]); G
+            Permutation Group with generators [(1,2)(3,4)]
+            sage: pi = {1: [1,2], 2: [3,4]}
+            sage: E2XY = At(G, pi)
+            sage: At(G, pi).rename("E_2(XY)"); E2XY
+            E_2(XY)
+            sage: m = P._indices.gen((G, pi)); m
```

Comment:
This looks a bit fishy to me: why should `gen` accept anything but an atomic species?  It seems to me that only `_element_constructor_ ` should accept input that is not in the right form.

I think you could remove `gen`, and make `_project` return an atomic species, no?

I now see that you then get an infinite recursion because of the group computation - can you remind me why it is better to compute the group always, rather than only when it is needed?

Reply:
My intention originally was to only compute the group when needed. It appears that I may have instead achieved the opposite, because I was trying to cache intermediate results for optimisation 🤔 ...